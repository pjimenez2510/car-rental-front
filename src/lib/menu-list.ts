import { UserRole } from "@/features/users/interfaces/user.interface";
import { Car, LucideIcon } from "lucide-react";

type Submenu = {
  href: string;
  label: string;
  active: boolean;
  roles: UserRole[];
};

type Menu = {
  href: string;
  label: string;
  active: boolean;
  roles: UserRole[];
  icon: LucideIcon;
  submenus: Submenu[];
};

type Group = {
  groupLabel: string;
  menus: Menu[];
};

const getAllMenuList = (pathname: string) => {
  const allMenus: Group[] = [
    {
      groupLabel: "MÃ³dulos",
      menus: [
        {
          href: "/management/car",
          label: "Autos",
          active: pathname.startsWith("/management/car"),
          roles: [UserRole.Admin, UserRole.Employee, UserRole.Customer],
          icon: Car,
          submenus: [
            {
              href: "/management/car/list",
              label: "Listar",
              active: pathname.startsWith("/management/car/list"),
              roles: [UserRole.Admin, UserRole.Employee],
            },
            {
              href: "/management/car/create",
              label: "Crear",
              active: pathname.startsWith("/management/car/create"),
              roles: [UserRole.Admin, UserRole.Employee],
            },
            {
              href: "/management/car/maintenances",
              label: "Mantenimientos",
              active: pathname.startsWith("/management/car/maintenances"),
              roles: [UserRole.Admin, UserRole.Employee],
            },
          ],
        },
      ],
    },
  ];
  return allMenus;
};

export const getMenuList = (pathname: string, role?: UserRole): Group[] => {
  if (!role) return [];

  const allMenus = getAllMenuList(pathname);

  return allMenus
    .map((group) => ({
      ...group,
      menus: group.menus
        .filter((menu) => menu.roles.includes(role))
        .map((menu) => ({
          ...menu,
          submenus: menu.submenus.filter((submenu) =>
            submenu.roles.includes(role)
          ),
        })),
    }))
    .filter((group) => group.menus.length > 0);
};
