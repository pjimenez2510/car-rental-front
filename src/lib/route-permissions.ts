import { UserRole } from "@/features/users/interfaces/user.interface";
import { pathToRegexp } from "path-to-regexp";

interface RoutesPrivate {
  path: string;
  roles: UserRole[];
}

const routes: RoutesPrivate[] = [
  {
    path: "/management/vehicles",
    roles: [UserRole.Admin, UserRole.Employee],
  },
  {
    path: "/management/vehicles/list",
    roles: [UserRole.Admin, UserRole.Employee],
  },
  {
    path: "/management/vehicles/create",
    roles: [UserRole.Admin, UserRole.Employee],
  },
  {
    path: "/management/vehicles/edit/:id",
    roles: [UserRole.Admin, UserRole.Employee],
  },
  {
    path: "/management/vehicles/maintenance/:id",
    roles: [UserRole.Admin, UserRole.Employee],
  },
  {
    path: "/reservations",
    roles: [UserRole.Customer],
  },
  {
    path: "/reservations/:id",
    roles: [UserRole.Customer],
  },
  {
    path: "/reservations/payment/:id",
    roles: [UserRole.Customer],
  },
  {
    path: "/new-reservation/:id",
    roles: [UserRole.Customer],
  },
];

interface RoleAllowed {
  path: boolean;
  role: boolean;
}

export function isRoleAllowed(path: string, role?: string): RoleAllowed {
  const route = routes.find((route) => {
    const regex = pathToRegexp(route.path);
    return regex.regexp.test(path);
  });

  if (!route) return { path: false, role: false };

  return { path: true, role: route.roles.includes(role as UserRole) };
}
