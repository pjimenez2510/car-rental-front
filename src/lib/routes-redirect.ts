import { UserRole } from "@/features/users/models/user.model";

export const routesRedirectAuth: Record<UserRole, string> = {
  admin: "/management/car",
  employee: "/management/car",
  customer: "/reservation",
};
