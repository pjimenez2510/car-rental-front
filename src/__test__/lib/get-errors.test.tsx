import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { getErrors } from "@/lib/get-errors";

describe("getErrors", () => {
  test("should return undefined when errors is undefined", () => {
    const result = getErrors(undefined);
    expect(result).toBeUndefined();
  });

  test("should render a single error message", () => {
    const singleError = "This is an error message";
    const { container } = render(getErrors(singleError));

    expect(container.querySelector("div")).toBeInTheDocument();
    expect(screen.getByText(singleError)).toBeInTheDocument();
    expect(container.querySelectorAll("p")).toHaveLength(1);
  });

  test("should render multiple error messages", () => {
    const multipleErrors = [
      "First error message",
      "Second error message",
      "Third error message",
    ];

    const { container } = render(getErrors(multipleErrors));

    expect(container.querySelector("div")).toBeInTheDocument();
    expect(container.querySelectorAll("p")).toHaveLength(3);

    multipleErrors.forEach((error) => {
      expect(screen.getByText(error)).toBeInTheDocument();
    });
  });

  test("should render nothing when empty array is provided", () => {
    const { container } = render(getErrors([]));

    expect(container.querySelector("div")).toBeInTheDocument();
    expect(container.querySelectorAll("p")).toHaveLength(0);
  });

  test("should render empty string correctly", () => {
    const { container } = render(getErrors(""));
    expect(container).toBeEmptyDOMElement();
  });

  test("should match snapshot with multiple errors", () => {
    const multipleErrors = ["First error message", "Second error message"];

    const { container } = render(getErrors(multipleErrors));
    expect(container).toMatchSnapshot();
  });
});
