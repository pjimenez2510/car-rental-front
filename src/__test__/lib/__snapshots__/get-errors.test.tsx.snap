// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getErrors should match snapshot with multiple errors 1`] = `
<div>
  <div>
    <p>
      First error message
    </p>
    <p>
      Second error message
    </p>
  </div>
</div>
`;
